packages/core/lib/commands/compile.js

```diff
@@ -67,6 +72,10 @@ const command = {
     const Config = require("@truffle/config");
     const config = Config.detect(options);
 
+    config.specificFiles = (options._ || []).map(sourcePath =>
+      path.resolve(config.working_directory, sourcePath)
```

Comment:
(I modified the implementation to do this resolution here. @robertmagier originally had these in a dedicated `compile.specific` method in each of the packages)